modules/generator/generator.go

```diff
@@ -0,0 +1,305 @@
+	ringStore, err := kv.NewClient(
+		cfg.Ring.KVStore,
+		ring.GetCodec(),
+		kv.RegistererWithKVName(prometheus.WrapRegistererWithPrefix("cortex_", reg), "metrics-generator"),
+		log.Logger,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("create KV store client: %w", err)
+	}
+
+	lifecyclerCfg, err := cfg.Ring.toLifecyclerConfig()
+	if err != nil {
+		return nil, fmt.Errorf("invalid ring lifecycler config: %w", err)
+	}
+
+	// Define lifecycler delegates in reverse order (last to be called defined first because they're
+	// chained via "next delegate").
+	delegate := ring.BasicLifecyclerDelegate(g)
+	delegate = ring.NewLeaveOnStoppingDelegate(delegate, log.Logger)
+	delegate = ring.NewAutoForgetDelegate(ringAutoForgetUnhealthyPeriods*cfg.Ring.HeartbeatTimeout, delegate, log.Logger)
+
+	g.ringLifecycler, err = ring.NewBasicLifecycler(lifecyclerCfg, ringNameForServer, RingKey, ringStore, delegate, log.Logger, prometheus.WrapRegistererWithPrefix("cortex_", reg))
+	if err != nil {
+		return nil, fmt.Errorf("create ring lifecycler: %w", err)
+	}
+
+	ringCfg := cfg.Ring.ToRingConfig()
+	g.ring, err = ring.NewWithStoreClientAndStrategy(ringCfg, ringNameForServer, RingKey, ringStore, ring.NewIgnoreUnhealthyInstancesReplicationStrategy(), prometheus.WrapRegistererWithPrefix("cortex_", reg), log.Logger)
+	if err != nil {
+		return nil, fmt.Errorf("create ring client: %w", err)
+	}
+
+	// Remote write
+	remoteWriteMetrics := remotewrite.NewMetrics(reg)
+	g.appendableFactory = func(userID string) storage.Appendable {
+		return remotewrite.NewAppendable(&cfg.RemoteWrite, log.Logger, userID, remoteWriteMetrics)
+	}
+
+	g.Service = services.NewBasicService(g.starting, g.running, g.stopping)
+	return g, nil
+}
+
+func (g *Generator) starting(ctx context.Context) (err error) {
+	// In case this function will return error we want to unregister the instance
+	// from the ring. We do it ensuring dependencies are gracefully stopped if they
+	// were already started.
+	defer func() {
+		if err == nil || g.subservices == nil {
+			return
+		}
+
+		if stopErr := services.StopManagerAndAwaitStopped(context.Background(), g.subservices); stopErr != nil {
+			level.Error(log.Logger).Log("msg", "failed to gracefully stop metrics-generator dependencies", "err", stopErr)
+		}
+	}()
+
+	g.subservices, err = services.NewManager(g.ringLifecycler, g.ring)
+	if err != nil {
+		return fmt.Errorf("unable to start metrics-generator dependencies: %w", err)
+	}
+	g.subservicesWatcher = services.NewFailureWatcher()
+	g.subservicesWatcher.WatchManager(g.subservices)
+
+	err = services.StartManagerAndAwaitHealthy(ctx, g.subservices)
+	if err != nil {
+		return fmt.Errorf("unable to start mertics-generator dependencies: %w", err)
+	}
+
+	return nil
+}
+
+func (g *Generator) running(ctx context.Context) error {
+	collectMetricsTicker := time.NewTicker(g.cfg.CollectionInterval)
+	defer collectMetricsTicker.Stop()
+
+	for {
+		select {
+		case <-collectMetricsTicker.C:
+			g.collectMetrics()
```

Comment:
if the timeout on the RW client is > then the collection interval we could start to see some backup here. Should we force the client timeout to be equal to the collection interval?

Reply:
We enforce the deadline in `g.collectMetrics()`. So each tenant can use up to collection interval seconds. But agree this doesn't avoid backup issues entirely. I.e. if every tenant takes 10 seconds and we have 4 tenants this call would still hang for way too long.
I'll add a global context enforcing the deadline across all tenants.

We will also want to revisit this code when we add per-tenant collection intervals.

Reply:
Checking out this code a bit more, the current implementation already considers this. `g.collectMetrics` creates a context with a timeout of `g.cfg.CollectionInterval`. So if collecting metrics for all tenants exceeds the collection interval, we abort the operation.

Would it be clearer if I created this context in `running`? I added it in `collectMetrics` because then we can do `defer cancel()` to clean up resources.

Code: https://github.com/grafana/tempo/blob/4089b5299ac1ba6e4a70d7fe5eb10c83890067b9/modules/generator/generator.go#L253-L266

Reply:
I see now. Nope, I'm good as is. Let's go!